File: Project - ObjC/Project - ObjC/Items/Shape/Cylinder.m


#import "Cylinder.h"


@implementation Cylinder

+ (instancetype)cylinder {
    return [self create];
}

- (instancetype)init {
    if (self = [super init]) {
        self.cylinder = [SCNCylinder new];
    }
    return self;
}

- (instancetype)initWithRadius:(CGFloat)radius height:(CGFloat)height {
    if (self = [super init]) {
        self.cylinder = [SCNCylinder cylinderWithRadius:radius height:height];
    }
    return self;
}

////////////////////////////////////////////////////////////////////////////////
#pragma mark - Property Overriding

- (void)setCylinder:(SCNCylinder *)cylinder {
    self.geometry = cylinder;
}

- (SCNCylinder *)cylinder {
    return (SCNCylinder *)self.geometry;
}


- (void)setRadius:(CGFloat)radius {
    self.cylinder.radius = radius;
}

- (CGFloat)radius {
    return self.cylinder.radius;
}

- (void)setHeight:(CGFloat)height {
    self.cylinder.height = height;
}

- (CGFloat)height {
    return self.cylinder.height;
}

@end

File: Project/Project/Items/Shapes/Cylinder.swift

import Foundation
import SceneKit
import JavaScriptCore



@objc class Cylinder: Shape {
    
    var radius: CGFloat {
        set { cylinder.radius = newValue }
        get { return cylinder.radius     }
    }
    
    var height: CGFloat {
        set { cylinder.height = newValue }
        get { return cylinder.height     }
    }
    
    override var size: CGFloat {
        set {
            radius = newValue / 2
            height = newValue
        }
        get {
            return max(radius * 2, height)
        }
    }
    
    var cylinder: SCNCylinder {
        set { geometry = newValue            }
        get { return geometry as SCNCylinder }
    }
   
    override init() {
        super.init(geometry: SCNCylinder())
    }
}

